### DISCUSSION TITLE
Infer contract type from ABI

### DISCUSSION THREAD

**Discussion by iduuck** [wevm/wagmi#34] [Ideas]
I am thinking about a way to infer the contract interface by just supplying the ABI. So that the potential typescript user could easily use the IDE/LSP to autocomplete the function calls, etc.

**Comment by tmm** [comment]
This is on my radar!

ethers.js v6 (dropping in a few weeks at earliest) will include improved TS support (including generics support for `Contract`) and should make this trivial. Happy to consider interim solutions until then.

Here's [a quick sketch](https://gist.github.com/tmm/8dcc049e8605256cb04634ac5e180bbd) I did for typing ether's Contract and you can see `ethereal-react`'s approach [here](https://github.com/kesne/ethereal-react/blob/main/packages/ethereal-react/src/contracts.ts#L70-L73).

**Reply by iduuck** [reply]
I like your quick sketch. However, what I was thinking about was even inferring them, and not only implementing some sort of generics. I love to have clean code, where most of the stuff is (of course) correctly inferred. I will have a dig into this, and come back with my own quick sketch. But very nice first draft.

**Reply by tmm** [reply]
Sounds great! Looking forward to it.

**Reply by liho00** [reply]
+1

**Reply by javierlinked** [reply]
+1

**Reply by tmm** [reply]
Forgot to circle back here. We [shipped this in `0.7`](https://twitter.com/wagmi_sh/status/1580591788611338240) (powered by our new [ABIType](https://github.com/wagmi-dev/abitype) library)! Check out wagmi's [TypeScript docs](https://wagmi.sh/docs/typescript) for more info.
